# Stock Analysis and Back Testing a Stratergy

**Hypothesis:**

- Let Previous day close of 3:25 be x.
- Let Previous day open at 9:15 be y.
- Current day open at 9:15 be z.
- Let max(high) from 9:15 to 10:30 for current day be alpha.

Pre-requisite:
- x > y
- z < x
- Calculate alpha.

If prerequisite is satisified, find a tick which satisfies Buy condition.

Buy if:
- close of a tick > alpha
- close of a tick > sma50 for that tick
- close of a tick > dma50 for that day

Sell if:
- close value of any tick < sma50


**Strategy:**

Buy if:
- Previous day range is positive.
- Today is a gap down. 
- Breaks the high made upto 10.30 
- Breakout bar is above 50sma  in dtf and 5 min.

Stop loss at:
- Breakout candle low for today and if trade is carried over then 50 sma.

## Import Liabraries 

In [1]:
import pandas as pd
import numpy as np
import datetime as dt
from pandas_datareader import data as pdr
import xlsxwriter
import openpyxl
from statistics import mean

## Import Stock Data 

- The data is stock price of **Bank Nifty**.

- We have data of **Open, High, Low, Close, SMA20, SMA50, SMA100**.

- The data starts from **July 2011** and goes till **January 2022**.

- The data is present of every **5 minute** interval starting from **9:15 am** and goes upto **3:25 pm**. 

- In total there are **75 intervals** in a day.

In [76]:
nse_data = pd.read_excel("Bank Nifty data 5 min.xlsx", header = 2)

In [77]:
nse_data['Date'] = nse_data['Date'].dt.date

In [78]:
nse_data.iloc[0]["Date"]

datetime.date(2011, 7, 29)

In [79]:
nse_data.head()

Date      Time      Open      High      Low      Close  SMA20  SMA50  \
0  2011-07-29  09:15:00  10820.65  10820.65  10758.40  10774.20    NaN    NaN   
1  2011-07-29  09:20:00  10774.20  10812.40  10774.20  10807.95    NaN    NaN   
2  2011-07-29  09:25:00  10807.95  10837.25  10804.35  10837.25    NaN    NaN   
3  2011-07-29  09:30:00  10836.05  10891.45  10835.60  10884.80    NaN    NaN   
4  2011-07-29  09:35:00  10884.80  10892.30  10871.85  10891.65    NaN    NaN   

   SMA 100  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN

In [80]:
nse_data = nse_data.fillna(-1)
nse_data.head()

Date      Time      Open      High      Low      Close  SMA20  SMA50  \
0  2011-07-29  09:15:00  10820.65  10820.65  10758.40  10774.20   -1.0   -1.0   
1  2011-07-29  09:20:00  10774.20  10812.40  10774.20  10807.95   -1.0   -1.0   
2  2011-07-29  09:25:00  10807.95  10837.25  10804.35  10837.25   -1.0   -1.0   
3  2011-07-29  09:30:00  10836.05  10891.45  10835.60  10884.80   -1.0   -1.0   
4  2011-07-29  09:35:00  10884.80  10892.30  10871.85  10891.65   -1.0   -1.0   

   SMA 100  
0     -1.0  
1     -1.0  
2     -1.0  
3     -1.0  
4     -1.0

In [81]:
nse_data.tail()

Date      Time      Open      High      Low      Close  \
192328  2022-01-24  15:05:00  36874.30  36946.70  36813.45  36946.70   
192329  2022-01-24  15:10:00  36946.55  37063.65  36929.50  37038.35   
192330  2022-01-24  15:15:00  37048.60  37073.95  36990.90  37050.30   
192331  2022-01-24  15:20:00  37047.80  37062.85  36894.25  36905.65   
192332  2022-01-24  15:25:00  36902.40  36945.05  36808.15  36834.90   

             SMA20      SMA50       SMA 100  
192328  36725.6400  36968.661  37242.978218  
192329  36735.3350  36960.372  37237.150000  
192330  36747.5000  36953.933  37231.681188  
192331  36751.2025  36944.372  37223.801980  
192332  36751.3125  36934.724  37215.664356

In [82]:
nse_data.describe()

Open           High           Low           Close  \
count  192333.000000  192333.000000  192333.000000  192333.000000   
mean    20697.310681   20713.934408   20679.911383   20697.032113   
std      8219.052185    8224.819853    8212.829251    8218.886333   
min      7772.600000    7801.950000    7766.350000    7772.600000   
25%     12713.500000   12721.650000   12703.900000   12713.250000   
50%     19198.800000   19214.450000   19180.700000   19197.900000   
75%     26863.350000   26877.750000   26846.200000   26862.650000   
max     41729.900000   41827.650000   41600.550000   41722.200000   

               SMA20          SMA50        SMA 100  
count  192333.000000  192333.000000  192333.000000  
mean    20694.677328   20690.957457   20684.575638  
std      8220.118405    8222.057737    8225.221458  
min        -1.000000      -1.000000      -1.000000  
25%     12709.005000   12706.930000   12701.710396  
50%     19198.837500   19194.966000   19185.750990  
75%     26864.802500   26870.704000   26873.287624  
max     41556.880000   41367.579000   41231.838614

In [83]:
nse_data.iloc[0]["Date"].year

2011

## Import Secondary Data

In [84]:
secondary_data = pd.read_excel("Data BN.xlsm", usecols = ["Date", "50 Dma"])

In [85]:
secondary_data['Date'] = secondary_data['Date'].dt.date

In [86]:
secondary_data.head()

Date  50 Dma
0  2014-01-01     NaN
1  2014-01-02     NaN
2  2014-01-03     NaN
3  2014-01-06     NaN
4  2014-01-07     NaN

In [87]:
secondary_data.tail()

Date     50 Dma
2021  2022-03-07  37225.998
2022  2022-03-08  37192.019
2023  2022-03-09  37167.170
2024  2022-03-10  37153.006
2025  2022-03-11  37143.023

In [88]:
nse_data = pd.merge(nse_data, secondary_data, on=['Date'])

In [89]:
nse_data.head()

Date      Time      Open      High      Low      Close       SMA20  \
0  2014-01-01  09:15:00  11418.90  11432.55  11388.00  11388.00  11368.1750   
1  2014-01-01  09:20:00  11390.75  11390.75  11361.00  11375.60  11368.6725   
2  2014-01-01  09:25:00  11375.60  11391.55  11373.55  11391.55  11369.8850   
3  2014-01-01  09:30:00  11391.55  11392.95  11381.10  11385.00  11370.8225   
4  2014-01-01  09:35:00  11385.00  11388.85  11377.75  11377.75  11371.5300   

       SMA50       SMA 100  50 Dma  
0  11362.309  11365.808911     NaN  
1  11362.511  11365.739604     NaN  
2  11363.107  11365.784158     NaN  
3  11363.794  11365.911386     NaN  
4  11364.071  11365.981188     NaN

In [90]:
nse_data.tail()

Date      Time      Open      High      Low      Close  \
147419  2022-01-24  15:05:00  36874.30  36946.70  36813.45  36946.70   
147420  2022-01-24  15:10:00  36946.55  37063.65  36929.50  37038.35   
147421  2022-01-24  15:15:00  37048.60  37073.95  36990.90  37050.30   
147422  2022-01-24  15:20:00  37047.80  37062.85  36894.25  36905.65   
147423  2022-01-24  15:25:00  36902.40  36945.05  36808.15  36834.90   

             SMA20      SMA50       SMA 100     50 Dma  
147419  36725.6400  36968.661  37242.978218  36819.007  
147420  36735.3350  36960.372  37237.150000  36819.007  
147421  36747.5000  36953.933  37231.681188  36819.007  
147422  36751.2025  36944.372  37223.801980  36819.007  
147423  36751.3125  36934.724  37215.664356  36819.007

In [91]:
nse_data.sample(10)

Date      Time      Open      High      Low      Close  \
108276  2019-12-18  10:30:00  32051.70  32070.05  32050.80  32070.05   
92589   2019-02-06  09:20:00  27319.60  27377.45  27309.20  27361.45   
72821   2018-01-10  12:00:00  25552.65  25575.45  25549.50  25562.40   
138827  2021-08-09  11:35:00  36033.95  36062.55  36032.20  36052.80   
99997   2019-07-03  14:10:00  31449.35  31450.40  31411.40  31426.85   
45560   2016-07-21  15:05:00  18683.05  18694.95  18670.10  18690.95   
24081   2015-05-26  12:25:00  18305.95  18305.95  18288.85  18295.55   
145186  2021-12-14  10:15:00  36660.15  36660.15  36581.50  36585.00   
29866   2015-09-10  13:15:00  16430.25  16440.60  16410.50  16430.55   
22059   2015-04-16  12:40:00  18557.60  18557.60  18536.20  18548.25   

             SMA20      SMA50       SMA 100     50 Dma  
108276  32109.6400  32138.682  32130.121782  30533.150  
92589   27279.8775  27294.153  27233.000000  27049.496  
72821   25601.1400  25650.442  25663.475248  25464.617  
138827  36036.8825  35927.657  35895.044554  35151.171  
99997   31430.8850  31377.530  31319.742574  30407.604  
45560   18765.9425  18829.181  18894.778218  17708.717  
24081   18312.4875  18308.398  18358.122277  18373.898  
145186  36776.1825  36949.559  37164.741089  38237.333  
29866   16374.1425  16320.336  16495.786634  18124.196  
22059   18571.2150  18614.777  18799.377228  18917.988

In [92]:
nse_data.iloc[5000:5010]

Date      Time      Open      High      Low      Close  \
5000  2014-04-07  09:50:00  12591.95  12593.05  12580.15  12581.40   
5001  2014-04-07  09:55:00  12581.40  12585.85  12565.20  12571.90   
5002  2014-04-07  10:00:00  12571.90  12582.35  12568.05  12570.95   
5003  2014-04-07  10:05:00  12570.95  12578.90  12561.30  12576.25   
5004  2014-04-07  10:10:00  12576.25  12577.05  12562.95  12565.85   
5005  2014-04-07  10:15:00  12565.85  12567.00  12528.45  12528.45   
5006  2014-04-07  10:20:00  12531.95  12557.60  12531.95  12552.05   
5007  2014-04-07  10:25:00  12552.05  12552.60  12522.25  12525.25   
5008  2014-04-07  10:30:00  12525.25  12536.45  12514.45  12536.45   
5009  2014-04-07  10:35:00  12536.45  12538.15  12505.30  12516.55   

           SMA20      SMA50       SMA 100     50 Dma  
5000  12564.1550  12576.920  12558.912871  11289.266  
5001  12565.2350  12576.327  12559.021287  11289.266  
5002  12566.0375  12575.680  12559.120792  11289.266  
5003  12567.4925  12575.389  12559.247525  11289.266  
5004  12568.4625  12574.929  12559.362376  11289.266  
5005  12567.3200  12573.523  12559.054950  11289.266  
5006  12567.8100  12572.636  12558.823762  11289.266  
5007  12567.5925  12570.967  12558.376238  11289.266  
5008  12567.8175  12569.476  12558.132178  11289.266  
5009  12566.7500  12567.908  12558.038614  11289.266

In [95]:
nse_data.dropna(inplace=True)

In [96]:
nse_data.head()

Date      Time      Open      High      Low      Close  \
3675  2014-03-12  09:15:00  11997.85  12049.25  11984.80  11998.70   
3676  2014-03-12  09:20:00  11998.70  12031.40  11967.75  11967.75   
3677  2014-03-12  09:25:00  11968.05  11990.40  11949.45  11968.85   
3678  2014-03-12  09:30:00  11968.85  11968.85  11922.90  11938.55   
3679  2014-03-12  09:35:00  11935.80  11949.80  11928.65  11949.80   

           SMA20      SMA50       SMA 100     50 Dma  
3675  12033.6700  12064.246  12090.860891  10814.169  
3676  12028.0375  12061.526  12089.981683  10814.169  
3677  12023.4075  12058.600  12089.100990  10814.169  
3678  12017.5875  12054.983  12088.296535  10814.169  
3679  12013.0550  12051.252  12087.416832  10814.169

In [97]:
nse_data.tail()

Date      Time      Open      High      Low      Close  \
147419  2022-01-24  15:05:00  36874.30  36946.70  36813.45  36946.70   
147420  2022-01-24  15:10:00  36946.55  37063.65  36929.50  37038.35   
147421  2022-01-24  15:15:00  37048.60  37073.95  36990.90  37050.30   
147422  2022-01-24  15:20:00  37047.80  37062.85  36894.25  36905.65   
147423  2022-01-24  15:25:00  36902.40  36945.05  36808.15  36834.90   

             SMA20      SMA50       SMA 100     50 Dma  
147419  36725.6400  36968.661  37242.978218  36819.007  
147420  36735.3350  36960.372  37237.150000  36819.007  
147421  36747.5000  36953.933  37231.681188  36819.007  
147422  36751.2025  36944.372  37223.801980  36819.007  
147423  36751.3125  36934.724  37215.664356  36819.007

In [98]:
nse_data.reset_index(drop=True, inplace=True)

In [99]:
nse_data.head()

Date      Time      Open      High      Low      Close       SMA20  \
0  2014-03-12  09:15:00  11997.85  12049.25  11984.80  11998.70  12033.6700   
1  2014-03-12  09:20:00  11998.70  12031.40  11967.75  11967.75  12028.0375   
2  2014-03-12  09:25:00  11968.05  11990.40  11949.45  11968.85  12023.4075   
3  2014-03-12  09:30:00  11968.85  11968.85  11922.90  11938.55  12017.5875   
4  2014-03-12  09:35:00  11935.80  11949.80  11928.65  11949.80  12013.0550   

       SMA50       SMA 100     50 Dma  
0  12064.246  12090.860891  10814.169  
1  12061.526  12089.981683  10814.169  
2  12058.600  12089.100990  10814.169  
3  12054.983  12088.296535  10814.169  
4  12051.252  12087.416832  10814.169

In [109]:
nse_data.iloc[0+74]

Date         2014-03-12
Time           15:25:00
Open           12025.05
High            12040.8
Low            12023.05
Close           12040.8
SMA20         12010.765
SMA50         12020.582
SMA 100    12025.391089
50 Dma        10814.169
Name: 74, dtype: object

In [101]:
nse_data.to_excel("Bank Nifty with 50 DMA.xlsx", index=False)

In [103]:
wb = openpyxl.Workbook()

ws = wb.active
ws.title = "Summary"

ws.column_dimensions['F'].width = 30
ws.column_dimensions['G'].width = 8
ws["A1"].value = """
This excel sheet contains output data gathered from backtesting strategy 2.
"""
wb.save('Strategy 2.xlsx')

## 1. Find the number of days where it satisfies the prerequsite.

In [104]:
len(nse_data.index)

143749

In [222]:
nse_data.iloc[0,:]

Date         2014-03-12
Time           09:15:00
Open           11997.85
High           12049.25
Low             11984.8
Close           11998.7
SMA20          12033.67
SMA50         12064.246
SMA 100    12090.860891
50 Dma        10814.169
Name: 0, dtype: object

In [165]:
index = 0
max_high = max(nse_data.iloc[index:index+16]["High"])
print(max_high)

12142.9


In [178]:
[nse_data.loc[(nse_data['Time'] == dt.time(11,15,0)) & (nse_data['Date'] == dt.date(2015, 10, 12))]].index

<function list.index(value, start=0, stop=9223372036854775807, /)>

In [245]:
def num_days_buy_condition_true(stage):
    condition_satisfied = condition_not_satisfied = 0
    index = 0
    all_profits = []
    while index < len(nse_data.index):
        
        if stage == "prerequsite":
            if nse_data["Time"][index] == dt.time(9,15,0):
                previous_open = nse_data["Open"][index]
                index += 74 
                continue
            if nse_data["Time"][index] == dt.time(15,25,0):
                previous_close = nse_data["Close"][index]

                if previous_close > previous_open:
                    current_open = nse_data["Open"][index+1]
                    if current_open < previous_close:
                        stage = "buy"
                        previous_day = nse_data["Date"][index]
                        index += 1 
                        max_high = max(nse_data.iloc[index:index+16]["High"])
                        index += 16
                    else:
                        index += 1
                else:
                    index += 1
            else:
                if nse_data["Time"][index] != dt.time(9,15,0):
                    index += 1
            
            """
            Figure out what to do when in prereq. stage and tick not in start or end row.
            Index increment should be such that it goes to next days 9:15 tick.
            """      
        if stage == "buy":
            if all(nse_data["Close"][index] > x for x in [max_high, nse_data["SMA50"][index],nse_data["50 Dma"][index]]):
                buy_price = nse_data["Close"][index]
                buy_date = nse_data["Date"][index]
                buy_time = nse_data["Time"][index]
                stage = "sell"
            index += 1
                
        if stage == "sell":
            if nse_data["Close"][index] < nse_data["SMA50"][index]:
            #if nse_data["Time"][index] == dt.time(15,25,0):
                sell_price = nse_data["Close"][index] 
                sell_date = nse_data["Date"][index]
                sell_time = nse_data["Time"][index]
                stage = "prerequsite"
                index += 1
                current_profit = sell_price - buy_price
                all_profits.append(current_profit)
                #print(previous_open, previous_close, previous_day, max_high, buy_price, sell_price, sell_price-buy_price, buy_date, buy_time, sell_date, sell_time)
            else:
                index += 1
    return all_profits

In [246]:
all_profits = num_days_buy_condition_true("prerequsite")

In [247]:
len(all_profits)

105

In [249]:
sum(all_profits)

4986.20000000001

In [250]:
max(all_profits)

1344.7999999999956

In [251]:
min(all_profits)

-632.6500000000015

In [ ]:
condition_satisfied, condition_not_satisfied = num_days_buy_condition_true()
print("Number of days condition was satisfied: " + str(condition_satisfied))
print("Number of days condition was not satisfied: " + str(condition_not_satisfied))


ws["a2"] = "Number of days condition was satisfied: " + str(condition_satisfied)
ws["a3"] = "Number of days condition was not satisfied: " + str(condition_not_satisfied)
wb.save("strategy1.xlsx")

 ### Inference: This tells us that there were 146 days where the stock price of BANK NIFTY was on rise throughout the day.
 
 i.e. The price of stock at any point of time on that day was greater than the SMA50 (average of price of the last 50 points)

## 2. Find the number of days where all 75 ticks had close value less than SMA50 

In [ ]:
def num_days_stock_declining():
    condition_satisfied = condition_not_satisfied = 0

    for index in nse_data.index:
        if nse_data["Time"][index] == dt.time(9,15,0):
            counter = 0

        if nse_data["Close"][index] <= nse_data["SMA50"][index]:
            counter += 1
        else:
            counter = -1

        if nse_data["Time"][index] == dt.time(15,25,0):
            if counter == 75:
                # all 75 ticks satisfied the condition
                condition_satisfied += 1
            else:
                # condition failed
                condition_not_satisfied += 1
                
    return condition_satisfied, condition_not_satisfied

In [ ]:
condition_satisfied, condition_not_satisfied = num_days_stock_declining()
print("Number of days condition was satisfied: " + str(condition_satisfied))
print("Number of days condition was not satisfied: " + str(condition_not_satisfied))

 ### Inference: This tells us that there were 105 days where the stock price of BANK NIFTY was on fall throughout the day.
 
 i.e. The price of stock at any point of time on that day was lower than the SMA50 (average of price of the last 50 points)

## 3. Buy the Stock at the end of the day when the Price is at rice and Sell when it starts to decline.


Stratergy:
- Buy at 3:25 pm on the day when all 75 ticks had close value greater than the SMA50.
- Sell at the first instant when the close value is less than the SMA50.
- Find out the money made or lost with this stratergy over last 10 year.



In [ ]:
def execute_trade(start_index, end_index):
    """
    End Index in not inclusive.
    
    """
    buying_mode = True
    profit_made_sma50 = []
    trade_number = 1
    trades = {}
    year = nse_data.iloc[start_index]["Date"].year

    for index in range(start_index, end_index):
        if buying_mode == True:    
            if nse_data["Time"][index] == dt.time(9,15,0):
                counter = 0

            if nse_data["Close"][index] >= nse_data["SMA50"][index]:
                counter += 1
            else:
                counter = -1

            if nse_data["Time"][index] == dt.time(15,25,0):
                if counter == 75:
                    # all 75 ticks satisfied the condition
                    # trigger buy
                    buy_price = nse_data["Close"][index]
                    buying_mode = False
                    trades[f"Trade {trade_number} Buy"] = nse_data.iloc[index:index+1,[0,1,5,7]]
                    continue

        if buying_mode == False:
            #print(nse_data.iloc[index:index+1,[0,1, 5,7]])
            if nse_data["Close"][index] < nse_data["SMA50"][index]:
                trades[f"Trade {trade_number} Sell"] = nse_data.iloc[index:index+1,[0,1,5,7]]
                selling_price = nse_data["Close"][index]
                profit_made = selling_price - buy_price
                trades[f"Trade {trade_number} Profit"] = profit_made
                trade_number += 1
                profit_made_sma50.append(profit_made)
                buying_mode = True
    return (profit_made_sma50, trades, year)

## 4. Find number of years:

In [ ]:
def save_output_parameters(profit_list, trades, year):
    win_trades = list(filter(lambda x: (x > 0), profit_list))
    loose_trades = list(filter(lambda x: (x < 0), profit_list))
    global cumulative_profit
    cumulative_profit.extend(profit_list)
    
    try:
        ws = wb[str(year)]
    except:
        wb.create_sheet(str(year))
        ws = wb[str(year)]
        ws.title = str(year)
    ws.column_dimensions['F'].width = 30
    ws.column_dimensions['G'].width = 8
    
    ws['f11'], ws['g11'] = "Total Trades", len(profit_list)
    ws['f12'], ws['g12'] = "Total Win Trades", len(win_trades)
    ws['f13'], ws['g13'] = "Total Loose Trades", len(loose_trades)
    ws['f14'], ws['g14'] = "Hit Ratio/ Trade win Ratio", float(f'{len(win_trades)/len(profit_list)*100:.2f}') if len(profit_list) != 0 else "NA"
    ws['f15'], ws['g15'] = "Day win Ratio", float(f'{len(win_trades)/len(loose_trades)*100:.2f}') if len(loose_trades) != 0 else "NA"
    ws['f16'], ws['g16'] = "Gross Profit", sum(win_trades)
    ws['f17'], ws['g17'] = "Gross Loss", sum(loose_trades)
    ws['f18'], ws['g18'] = "Net Profit", sum(win_trades)+sum(loose_trades)
    ws['f19'], ws['g19'] = "Max Draw Up", "???"
    ws['f20'], ws['g20'] = "Max Draw Down", "???"
    ws['f21'], ws['g21'] = "Average Trades", float(f"{ws['g18'].value/len(profit_list):.2f}")  if len(profit_list) != 0 else "NA"
    ws['f22'], ws['g22'] = "Average Winning Trade", float(f'{mean(win_trades):.2f}') if len(win_trades) != 0 else "NA"
    ws['f23'], ws['g23'] = "Average Loosing Trade", float(f'{mean(loose_trades):.2f}') if len(loose_trades) != 0 else "NA"
    ws['f24'], ws['g24'] = "Risk Reward", "???"
    ws['f25'], ws['g25'] = "Max Return/ Largest Winning Trade", max(win_trades) if len(win_trades) != 0 else "NA"
    ws['f26'], ws['g26'] = "Min Return/ Largest Loosing Trade", min(loose_trades) if len(loose_trades) != 0 else "NA"
    ws['f27'], ws['g27'] = "Net Profit %", "???"
    ws['f28'], ws['g28'] = "Return Per Year", ws['g18'].value
    ws['f29'], ws['g29'] = "Sharpe Ratio", "???"
    wb.save("strategy1.xlsx")

In [ ]:
last_index = 0
last_year = nse_data.iloc[0]["Date"].year

year_start_index_number = [last_index]

for index in nse_data.index:
    if nse_data.iloc[index]["Date"].year > last_year:
        last_year = nse_data.iloc[index]["Date"].year
        year_start_index_number.append(index)

In [ ]:
global cumulative_profit
cumulative_profit = []
for i in range(len(year_start_index_number)):
    start_index = year_start_index_number[i]
    if i < len(year_start_index_number)-1:
        end_index = year_start_index_number[i+1]
    profit_made, trades, year = execute_trade(start_index, end_index)
    save_output_parameters(profit_made, trades, year)

In [ ]:
len(cumulative_profit)

In [ ]:
win_trades = list(filter(lambda x: (x > 0), cumulative_profit))
loose_trades = list(filter(lambda x: (x < 0), cumulative_profit))

ws = wb["Summary"]

ws['f11'], ws['g11'] = "Total Trades", len(cumulative_profit)
ws['f12'], ws['g12'] = "Total Win Trades", len(win_trades)
ws['f13'], ws['g13'] = "Total Loose Trades", len(loose_trades)
ws['f14'], ws['g14'] = "Hit Ratio/ Trade win Ratio", float(f'{len(win_trades)/len(cumulative_profit)*100:.2f}') if len(cumulative_profit) != 0 else "NA"
ws['f15'], ws['g15'] = "Day win Ratio", float(f'{len(win_trades)/len(loose_trades)*100:.2f}') if len(loose_trades) != 0 else "NA"
ws['f16'], ws['g16'] = "Gross Profit", sum(win_trades)
ws['f17'], ws['g17'] = "Gross Loss", sum(loose_trades)
ws['f18'], ws['g18'] = "Net Profit", sum(win_trades)+sum(loose_trades)
ws['f19'], ws['g19'] = "Max Draw Up", "???"
ws['f20'], ws['g20'] = "Max Draw Down", "???"
ws['f21'], ws['g21'] = "Average Trades", float(f"{ws['g18'].value/len(cumulative_profit):.2f}") if len(cumulative_profit) != 0 else "NA"
ws['f22'], ws['g22'] = "Average Winning Trade", float(f'{mean(win_trades):.2f}') if len(win_trades) != 0 else "NA"
ws['f23'], ws['g23'] = "Average Loosing Trade", float(f'{mean(loose_trades):.2f}') if len(loose_trades) != 0 else "NA"
ws['f24'], ws['g24'] = "Risk Reward", "???"
ws['f25'], ws['g25'] = "Max Return/ Largest Winning Trade", max(win_trades) if len(win_trades) != 0 else "NA"
ws['f26'], ws['g26'] = "Min Return/ Largest Loosing Trade", min(loose_trades) if len(loose_trades) != 0 else "NA"
ws['f27'], ws['g27'] = "Net Profit %", "???"
ws['f28'], ws['g28'] = "Sharpe Ratio", "???"
wb.save("strategy1.xlsx")